New .NET AI chat app template quickstart (#44988)
dotnet/docs: docs/ai/quickstarts/quickstart-visual-studio-ai-templates.md

* Access to an [Azure OpenAI service](/azure/ai-services/openai/how-to/provisioned-get-started) with the following configurations:
  - A `gpt-4o-mini` model deployed
  - A `text-embedding-3-small` model deployed
  - The [Azure AI Developer](/azure/role-based-access-control/built-in-roles/ai-machine-learning#azure-ai-developer) role assigned to the user you used to sign-in to Visual Studio or the Azure CLI with locally

:::zone-end

:::zone target="docs" pivot="openai"

* .NET 9.0 SDK - [Install the .NET 9.0 SDK](https://dotnet.microsoft.com/download)
* Visual Studio 2022 - [Install Visual Studio 2022](https://code.visualstudio.com/) (optional)
* Access to an [OpenAI service](https://openai.com/api/) with the following configurations:

:::zone-end

:::zone target="docs" pivot="ollama"

* .NET 9.0 SDK - [Install the .NET 9.0 SDK](https://dotnet.microsoft.com/download)
* Visual Studio 2022 - [Install Visual Studio 2022](https://code.visualstudio.com/) (optional)
* Ollama installed locally - [Install Ollama](https://ollama.com/) locally on your device

:::zone-end

## Install the .NET AI app template

> The **AI Chat with Custom Data** template is available as a template package through NuGet. Use the `dotnet new` command to install  the package:

Reviewer: ```suggestion
The **AI Chat with Custom Data** template is available as a template package through NuGet. Use the [`dotnet new`](../../core/tools/dotnet-new-install.md) command to install the package:
```

> The **AI Chat with Custom Data** template is available as a template package through NuGet. Use the `dotnet new` command to install  the package:

Reviewer 2: ```suggestion
The **AI Chat Web App** template is available as a template package through NuGet. Use the `dotnet new` command to install  the package:
```
We're changing the name for Preview 1 to clarify its use case.

> The **AI Chat with Custom Data** template is available as a template package through NuGet. Use the [`dotnet new`](../../core/tools/dotnet-new-install.md) command to install the package:

Author: ```suggestion
The **AI Chat Web App** template is available as a template package through NuGet. Use the [`dotnet new`](../../core/tools/dotnet-new-install.md) command to install the package:
```

```dotnetcli

> dotnet new --install AIChatCustomData

Reviewer 2: ```suggestion
dotnet new --install Microsoft.Extensions.AI.Templates
```

```

## Create the .NET AI app

> After you have installed the new AI app templates, the template is available through the Visual Studio UI or the .NET CLI.

Reviewer 2: We'll also support  VS Code with the C# Dev Kit. Trying to figure out some issues there now.

:::zone target="docs" pivot="azure-openai"

# [Visual Studio](#tab/visual-studio)

> 1. Inside Visual Studio, navigate to **File -> New -> Project**.

Reviewer: ```suggestion
1. Inside Visual Studio, navigate to **File > New > Project**.
```

> 1. On the **Create a new project** screen, search for **AI Chat with Custom Data**. Select the matching result and then choose **Next**.

Reviewer 2: ```suggestion
1. On the **Create a new project** screen, search for **AI Chat Web App**. Select the matching result and then choose **Next**.
```

1. On the **Configure your new project** screen, enter the desired name and location for your project and then choose **Next**.
1. On the **Additional information** screen:
    - For the **Framework** option, select **.NET 9.0**.
    - For the **AI service provider** option, select **Azure OpenAI**.
    - Make sure the **Use managed identity** checkbox is checked.
    - For the **Vector store** option, select **Local on-disc (for prototyping)**.
1. Select **Create** to complete the process.

# [.NET CLI](#tab/dotnet-cli)

1. In a terminal window, navigate to an empty directory on your device.
1. Create a new app with the `dotnet new` command and the following parameters:

    ```dotnetcli

>     dotnet new chat --framework "net9.0" --AiServiceProvider "azureopenai" --VectorStore "local"

Reviewer 2: ```suggestion
    dotnet new aichatweb --framework "net9.0" --AIServiceProvider "azureopenai" --VectorStore "local"
```

    ```

>     The .NET CLI creates a new .NET 9.0 app with the configurations your specified.

Reviewer: ```suggestion
    The .NET CLI creates a new .NET 9.0 app with the configurations you specified.
```

1. Open the new app in your editor of choice, such as Visual Studio Code.

    ```dotnetcli
    code .
    ```

---

:::zone-end

:::zone target="docs" pivot="openai"

# [Visual Studio](#tab/visual-studio)

> 1. Inside Visual Studio, navigate to **File -> New -> Project**.

Reviewer: ```suggestion
1. Inside Visual Studio, navigate to **File > New > Project**.
```

> 1. On the **Create a new project** screen, search for **AI Chat with Custom Data**. Select the matching result and then choose **Next**.

Reviewer 2: ```suggestion
1. On the **Create a new project** screen, search for **AI Chat Web App**. Select the matching result and then choose **Next**.
```

1. On the **Configure your new project** screen, enter the desired name and location for your project and then choose **Next**.
1. On the **Additional information** screen:
    - For the **Framework** option, select **.NET 9.0**.
    - For the **AI service provider** option, select **OpenAI**.

>     - For the **Vector store** option, select **Local on-disc (for prototyping)**.

Reviewer 2: ```suggestion
    - For the **Vector store** option, select **Local on-disk (for prototyping)**.
```

1. Select **Create** to complete the process.

# [.NET CLI](#tab/dotnet-cli)

1. In a terminal window, navigate to an empty directory on your device.
1. Create a new app with the `dotnet new` command and the following parameters:

    ```dotnetcli

>     dotnet new chat --framework "net9.0" --AiServiceProvider "openai" --VectorStore "local"

Reviewer 2: ```suggestion
    dotnet new aichatweb --framework "net9.0" --AIServiceProvider "openai" --VectorStore "local"
```

    ```

>     The .NET CLI creates a new .NET 9.0 app with the configurations your specified.

Reviewer: ```suggestion
    The .NET CLI creates a new .NET 9.0 app with the configurations you specified.
```

1. Open the new app in your editor of choice, such as Visual Studio Code.

    ```dotnetcli
    code .
    ```

---

:::zone-end

:::zone target="docs" pivot="ollama"

# [Visual Studio](#tab/visual-studio)

> 1. Inside Visual Studio, navigate to **File -> New -> Project**.

Reviewer: ```suggestion
1. Inside Visual Studio, navigate to **File > New > Project**.
```

> 1. On the **Create a new project** screen, search for **AI Chat with Custom Data**. Select the matching result and then choose **Next**.

Reviewer 2: ```suggestion
1. On the **Create a new project** screen, search for **AI Chat Web App**. Select the matching result and then choose **Next**.
```

1. On the **Configure your new project** screen, enter the desired name and location for your project and then choose **Next**.
1. On the **Additional information** screen:
    - For the **Framework** option, select **.NET 9.0**.
    - For the **AI service provider** option, select **Ollama**.

>     - For the **Vector store** option, select **Local on-disc (for prototyping)**.

Reviewer 2: ```suggestion
    - For the **Vector store** option, select **Local on-disk (for prototyping)**.
```

1. Select **Create** to complete the process.

# [.NET CLI](#tab/dotnet-cli)

1. In a terminal window, navigate to an empty directory on your device.
1. Create a new app with the `dotnet new` command and the following parameters:

    ```dotnetcli

>     dotnet new chat --framework "net9.0" --AiServiceProvider "ollama" --VectorStore "local"

Reviewer 2: ```suggestion
    dotnet new aichatweb --framework "net9.0" --AIServiceProvider "ollama" --VectorStore "local"
```

    ```

>     The .NET CLI creates a new .NET 9.0 app with the configurations your specified.

Reviewer: ```suggestion
    The .NET CLI creates a new .NET 9.0 app with the configurations you specified.
```

1. Open the new app in your editor of choice, such as Visual Studio Code.

    ```dotnetcli
    code .
    ```

---

:::zone-end

### Explore the sample app

The sample app you created is a Blazor Interactive Server web app preconfigured with common AI and data services. The app handles the following concerns for you:

- Includes essential `Microsoft.Extensions.AI` packages and other dependencies in the `csproj` file to help you get started working with AI.
- Creates various AI services and registers them for dependency injection in the `Program.cs` file:
  - An `IChatClient` service to chat back and forth with the generative AI model

>   - An `IEmbeddingGenerator` service that is used to generate embeddings, which are essential for vector search functionality

Reviewer: ```suggestion
  - An `IEmbeddingGenerator` service that's used to generate embeddings, which are essential for vector search functionality
```

  - A `JsonVectorStore` to act as an in-memory vector store
- Registers a SQLite database context service to handle ingesting documents. The app is preconfigured to ingest whatever documents you add to the `Data` folder of the project, including the provided `Example.pdf` file.
- Provides a complete chat UI using Blazor components. The UI handles rich formatting for the AI responses and provides features such as citations for response data.

## Configure the app

:::zone target="docs" pivot="azure-openai"

> The **AI Chat with Custom Data** app is almost ready to go as soon as it's created. However, you'll need to provide the endpoint for your Azure OpenAI service for the app to connect to. By default, the app template searches for this value in the project's local .NET user secrets.

Reviewer 2: ```suggestion
The **AI Chat Web App** app is almost ready to go as soon as it's created. However, you'll need to provide the endpoint for your Azure OpenAI service for the app to connect to. By default, the app template searches for this value in the project's local .NET user secrets.
```

> 1. Create a local .NET user secret to store the Azure OpenAI service endpoint:

Reviewer 2: We should include links to instructions for setting up the Azure OpenAI resources, as well as Azure AI Search if using that vector store.
